CMakeLists.txt

```diff
@@ -27,27 +30,118 @@ if (Git_FOUND)
   endif()
 endif()
 
+macro(set_cxx_flag FLAG)
+	if (${ARGC} GREATER 1)
+		set(CMAKE_CXX_FLAGS_${ARGV1} "${CMAKE_CXX_FLAGS_${ARGV1}} ${FLAG}")
+	else()
+		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${FLAG}")
+	endif()
+endmacro()
+
+macro(set_linker_flag FLAG)
+	if (${ARGC} GREATER 1)
+		set(CMAKE_EXE_LINKER_FLAGS_${ARGV1} "${CMAKE_EXE_LINKER_FLAGS_${ARGV1}} ${FLAG}")
+		set(CMAKE_SHARED_LINKER_FLAGS_${ARGV1} "${CMAKE_SHARED_LINKER_FLAGS_${ARGV1}} ${FLAG}")
+		set(CMAKE_MODULE_LINKER_FLAGS_${ARGV1} "${CMAKE_MODULE_LINKER_FLAGS_${ARGV1}} ${FLAG}")
+	else()
+		set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${FLAG}")
+		set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${FLAG}")
+		set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} ${FLAG}")
+	endif()
+endmacro()
+
+# This option decides if crunch is dynamically linked against libcrn.so
+# statically linked against libcrn.o, enabling it always build libcrn.so.
+# This option is a builtin CMake one, the name means “build executables
+# against shader libraries”, not “build the shared libraries”.
 option(BUILD_SHARED_LIBS "Link executables against shared library" OFF)
+# Always build libcrn.so even if crunch is linked to libcrn statically.
 option(BUILD_SHARED_LIBCRN "Build shared libcrn" OFF)
+# Always build libcrn.a even if crunch is linked to libcrn dynamically.
 option(BUILD_STATIC_LIBCRN "Build static libcrn" OFF)
+# Build the crunch tool, implies the build of libcrn.o or libcrn.so.
 option(BUILD_CRUNCH "Build crunch" ON)
+# Build the provided examples, they only build on Windows for now.
 option(BUILD_EXAMPLES "Build examples" OFF)
-if (NOT MSVC)
-  option(OPTIMIZE_RELEASE "Optimize release build" ON)
-endif()
+# Enable extra optimization flags, like using -O3 even in RelWithDebInfo build.
+option(USE_EXTRA_OPTIMIZATION "Enable extra optimization" ON)
+# Enable link time optimization, slows down the build but produce faster and smaller binaries.
+option(USE_LTO "Enable link-time optimization" ON)
+# Fast math can be used if generated files are written to a cache, but it is not a good idea
+# to use fast math if the generated files are meant to be distributed and reproducible.
+# See https://github.com/DaemonEngine/crunch/issues/29
+option(USE_FAST_MATH "Enable fast math (files generated with the executable may not be reproducible)" ON)
 
-set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR})
+if (MSVC)
+	# Enable MSVC parallel compilation.
+	set_cxx_flag("/MP")
+	
+	# Increase warning verbosity level.
+	set_cxx_flag("/W4")
 
-if (NOT MSVC)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wno-unused-value -Wno-unused")
-endif()
+	# At least Ninja doesn't remove the /W3 flag when we add /W4 or  /Wall one, which
+	# leads to compilation warnings. Remove /W3 entirely, as /W4 will be used.
+	foreach(flag_var
```

Comment:
I don't thing it's good to add this hacky flag replacing code. This stuff makes sense for Daemon as we enforce warning error on the CI and heavily customize the warning flags to remove ones that are too noisy. For Crunch there is no compelling reason to do this and it will just spam everyone with warnings about trivialities.

Reply:
Removed the foreach loop.

Reply:
OK, I removed `/W4` and `-Wall`.